### PR TITLE
feat: Enable GCS, S3, R2 and MinIO as object stores for local runs

### LINKED ISSUE #1806: Allow using GCS and S3 for underlying data storage when running locally
## Description

Currently we have the `--data-dir` flag which allows specifying a local path. We can extend this (or add another flag) to allow providing an object store bucket instead. This would act similarly to the existing `data-dir` behavior in that both the catalog and table data will be stored in the same place.

We also do some parsing the python lib to detect cloud vs. local. We can extend that to also detect s3/gs: https://github.com/glaredb/glaredb/blob/fb8d15c7d86c32d368503a42b27d3f95cc897410/py-glaredb/src/lib.rs#L60-L80

I'm not sure how we want to provide credentials here yet. Ideally we could look at the configured credentials in the user's home, but I'd want a way to manually specify the creds to. For python, we could use kwargs to independently specify access key, region, etc.

### PR DESCRIPTION
**Pull request by gruuya**
This PR introduces support for using external object stores as data/catalog storage in locally run GlareDB.

Still a couple of things missing:
- [ ] Test coverage
- [ ] Read creds from user's home

Closes #1806, and also partially addresses #1818.

### CODE REVIEW

**Comment by gruuya** [comment]
Verified it works on GCS:
```bash
$ export GOOGLE_SERVICE_ACCOUNT=/Users/markogrujic/GlareDB/glaredb-dev-playground-99fdd8e83a7c.json
$ cargo run --bin glaredb -- -l gs://gruuya-data-store-test
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /Users/markogrujic/GlareDB/glaredb/py-glaredb/Cargo.toml
workspace: /Users/markogrujic/GlareDB/glaredb/Cargo.toml
    Finished dev [unoptimized + debuginfo] target(s) in 0.66s
     Running `target/debug/glaredb -l 'gs://gruuya-data-store-test'`
GlareDB (v0.5.0)
Using in-memory catalog
Type \help for help.
> create table test as values (1, 'one'), (2, 'two');
Table created
> select * from test where column1 = 2;
┌─────────┬─────────┐
│ column1 │ column2 │
│      ── │ ──      │
│   Int64 │ Utf8    │
╞═════════╪═════════╡
│       2 │ two     │
└─────────┴─────────┘
```
Resulting delta table logs:
image

As for the S3 family of object stores there's a matter of deciding what to do with atomic copy-if-not-exists capabilities required for leasing. Cloudflare R2 has a header that enables this action: https://docs.rs/object_store/latest/object_store/aws/enum.S3CopyIfNotExists.html

For others (basically Amazon proper and MinIO) I see the options (not necessarily exclusive) as:
- do a non-atomic rename in two steps (check if the lease doesn't exist and then rename); this would leave us vulnerable to races, which would need to be communicated to the users in this case
- enable configuring an S3 lock client (like in delta-rs); this would provide the needed atomicity, though it would introduce additional complexity

File: crates/glaredb/src/args/mod.rs
Comment (vrongmeal) [reply crates/glaredb/src/args/mod.rs:104]: I like this. But let's rename `options` to `storage_options` or maybe move `--options` outside (in a more global scope).

File: crates/metastore/src/local.rs
Comment (vrongmeal) [inline crates/metastore/src/local.rs:31]: Nice! 🫣

File: crates/sqlexec/src/engine.rs
Comment (vrongmeal) [reply crates/sqlexec/src/engine.rs:44]: > If that was the case there would be no need for try_from_options logic, though some kind of validation logic would still be needed.

This should be fine. We do want something concrete after validation.

> Also needed would be the logic for hydrating the creds from the user's home if not supplied (the object store builders know how to read appropriate env vars on their own).

This would be a good addition!

Comment (gruuya) [reply crates/sqlexec/src/engine.rs:44]: > This should be fine. We do want something concrete after validation.

Thanks. Just to clarify, by concrete do you mean a single map with all the validated options or enum variants with explicit struct-like fields (as is done currently)? Basically trying to determine which one you'd prefer.

Comment (vrongmeal) [reply crates/sqlexec/src/engine.rs:44]: > Just to clarify, by concrete do you mean a single map with all the validated options or enum variants with explicit struct-like fields (as is done currently)? Basically trying to determine which one you'd prefer.

The latter (as current implementation).

Comment (gruuya) [reply crates/sqlexec/src/engine.rs:44]: Got it, will leave that part as is then, thanks!

**Comment by scsmithr** [comment]
Not requiring `copy_if_not_exists` would also help with https://github.com/GlareDB/glaredb/issues/1809

**Comment by tychoish** [comment]
> For others (basically Amazon proper and MinIO) I see the options (not necessarily exclusive) as:
> 
> * do a non-atomic rename in two steps (check if the lease doesn't exist and then rename); this would leave us vulnerable to races, which would need to be clearly communicated to the users
> * enable configuring an S3 lock client (like in delta-rs); this would provide the needed atomicity, though it would introduce additional complexity (particularly in the CLI, might as well implement a glare config file at that point)

Is there a third option that's "indirect [some?] names (or append random keys when writing to avoid clashes?)" 

Is there a fourth option that's do naming/renaming (again of some level?) as a (relatively simple) two phase commit?

----

I think we have to avoid clashing with ourselves more than anything: I don't think we need to solve "how do we avoid breaking things/being broken when someone manually moves files around while we're writing data. 

Am I misunderstanding things?

File: crates/object_store_util/src/shared.rs
Comment (gruuya) [inline crates/object_store_util/src/shared.rs:112]: > What I think we can do is change lease initialization logic to:
> 
>     1. Read lease. If it exists:
> 
> 
>     * return Ok if process id written in lease matches leaser's process id
> 
>     * return Err if process ids don't match
> 
> 
>     2. Put new lease (puts are atomic in s3)
> 
>     3. Read lease again to ensure no other process wrote the lease between steps 1 & 2.

In the end I realized that this alone wasn't going to work, since apart from us using `rename_if_not_exists` during lease initialization, and more importantly, `delta-rs` also uses it when committing a transaction. So I went with the above as a way to resolve both.

In this way, `NativeTableStorage` can pass a working object store to delta-rs, which can in turn commit the transactions. On the other hand the metastore client also gets initialized with a `SharedObjectStore` now, so the lease initialization will also work.

However, this does mean that races are still a potential problem, arguably more pronounced for transaction consistency than lease initialization. Therefore, in the long term we might need to implement the S3 lock client after all (or use `deltalake::s3::S3StorageBackend` which has one built in). 

One might also argue that this shouldn't be too big of a problem, since writes shouldn't be as frequent as reads (for OLAP workloads at least). Nonetheless, we could add a warning statement or an additional CLI switch so as to make the user fully aware of potential issues.

File: crates/metastore/Cargo.toml
Comment (vrongmeal) [inline crates/metastore/Cargo.toml:9]: Minor nitpick, but, we would want to add errors to the enum instead of using `anyhow` in this crate.

File: crates/object_store_util/Cargo.toml
Comment (vrongmeal) [inline crates/object_store_util/Cargo.toml:9]: Same comment as for `metastore`.

**Review by scsmithr — APPROVED** [review]

File: crates/sqlexec/src/engine.rs
Comment (gruuya) [reply crates/sqlexec/src/engine.rs:48]: > does it make sense to split into storage config into authentication components

Fwiw, I did try this initially; will try to change to that setup now as well.

> also does it make sense to have a "path prefix"

Good question; usually in these cases the bucket parameter can hold an arbitrary path as well, so I figured that would work, but after a quick test now I see it doesn't so I'll need to investigate/fix that.

File: crates/metastore/src/errors.rs
Comment (gruuya) [reply crates/metastore/src/errors.rs:92]: True, though I've seen this in various other error definitions in other crates so I figured this may be fine. Let me see if there's a another more refined alternative.

Comment (scsmithr) [reply crates/metastore/src/errors.rs:92]: For what it's worth, I've starting to come to the opinion that this is the easiest way to get a reasonable error. We currently have a lot of different error types and variants, and I originally approached it as "one failure case = one error variant", but this is tedious.

Also rust doesn't have backtraces for errors by default (and it's actually kinda painful to get them), and so if you use the same error variant twice in different spots, it becomes difficult to know which path caused that error.

File: crates/glaredb/src/args/mod.rs
Comment (gruuya) [reply crates/glaredb/src/args/mod.rs:97]: > what if you specify options which aren't supported

Unsupported options will be ignored but the initialization/setup will error out with a clear message (either from us, or courtesy of the object store crate itself) if some of the required keys are missing.

> or options that are specified more than once? will order matter (should we sort by key?)

The last option value will be used in this case.

File: crates/metastore/src/errors.rs
Comment (gruuya) [reply crates/metastore/src/errors.rs:92]: Yeah, this was my thinking too—too many error variants getting out of hand. 

Though in this case there is indeed a more suitable existing option for my use case that I missed (`MetastoreError::FailedInProcessStartup`), so I'll use that instead.

File: crates/object_store_util/src/shared.rs
Comment (tychoish) [reply crates/object_store_util/src/shared.rs]: > Therefore, in the long term we might need to implement the S3 lock client after all (or use deltalake::s3::S3StorageBackend which has one built in).

is there a downside to using the deltalake client? does it require changing some kind of "on-disk" format? should we just open an issue for that?

Comment (gruuya) [reply crates/object_store_util/src/shared.rs]: Not really, the only issue I see is the extra config needed to get the S3 lock working.

That said, I think I recall there were some plans of actually removing the delta-rs clients altogether: https://github.com/delta-io/delta-rs/issues/1521

File: crates/sqlexec/src/engine.rs
Comment (tychoish) [reply crates/sqlexec/src/engine.rs:47]: I think we should get that into main before cutting a release, but can merge this PR and then add prefixes in, if you want.

**Review by tychoish — APPROVED** [review]
(just for the record I think this is mergable modulo the changes discussed)

File: crates/sqlexec/src/engine.rs
Comment (gruuya) [reply crates/sqlexec/src/engine.rs:47]: Ok, will go ahead with merge, as I have other follow-up work as well (i.e. functional tests for this feature). Opened #1873 to track those two.

**Pull request merged**
